src/main/java/org/broadinstitute/barclay/argparser/CommandLinePluginDescriptor.java

```diff
@@ -76,103 +110,127 @@
+     * Implementations of this method should minimally validate that all values that have
      * been specified on the command line have a corresponding plugin instance (this will
      * detect a user-specified value for which there is no corresponding plugin class).
      *
-     * @throws CommandLineException if a plugin value has been specified that
-     * has no corresponding plugin instance (i.e., the plugin class corresponding to the name
-     * was not discovered)
+     * @throws CommandLineException if a plugin value has been specified that has no corresponding
+     * plugin instance (i.e., the plugin class corresponding to the name was not discovered)
+     * or the plugins otherwise cannot be resolved per the descriptor's argument resolution policy.
      */
-    public abstract void validateArguments() throws CommandLineException;
+    public abstract void validateAndResolvePlugins() throws CommandLineException;
 
     /**
-     * @return the default plugins used for this instance of this descriptor as a list of Object. Used for
-     * help/doc generation.
+     * @return the default plugins enabled for this descriptor as a list of Object. Used for
+     * help/documentation generation for this descriptor.
+     *
+     * NOTE: calling this method after argument parsing (and thus after {@link #validateAndResolvePlugins}
+     * is called) may return a different list than calling it before parsing, since the resolution
+     * policy is implementation-dependent and may depend on the actual argument specified by the user
+     * on the command line. For example, the user may have specified that a plugin originally included
+     * in the list of defaults be disabled. The corresponding plugin class would be returned by this
+     * mehtod BEFORE parsing, but not after, since it would have been removed by
+     * {@link #validateAndResolvePlugins}.
      */
-    public abstract List<Object> getDefaultInstances();
+    public abstract List<T> getDefaultInstances();
 
     /**
-     * @return an ordered List of actual plugin instances that have been specified on the command
-     * line, in the same order they were obtained/created by {@line #getInstance}).
+     * @return an ordered list containing the plugins that result from resolving all command line
+     * arguments with any default plugins that have been provided to this descriptor. This list
+     * represents the plugins that will actually be used by the consumer. The resolution policy is
+     * descriptor-dependent.
+     *
+     * NOTE: calling this method before argument parsing (and thus before {@link #validateAndResolvePlugins}
+     * has been called) may return a different list than calling it after parsing, since the resolution
+     * policy is implementation-dependent and may depend on the actual arguments specified by the user
+     * on the command line.
      */
-    public abstract List<T> getAllInstances();
```

Comment:
As the plugin specifies that all the instances should be kept in the descriptor, maybe it is a good idea to be able to have a getter for all the instances too (this will be useful for some purposes, like documentation).

Reply:
What would the use case, and contract, for that be ? What would "all" instances mean ? The reason we changed the name was to try to remove the ambiguity around that. Now we have getDefaultInstances, and getResolvedInstances, which have clear(er) meaning. I'm not sure how that would be used for doc ?

Reply:
Because the plugin is enforced to maintain the instances generated by `createInstanceForPlugin`, this will be a nice getter for retrieving all of them. Maybe instead of `getAllInstances`it should be named `getCreatedInstances`.

For documentation, it might be useful to get all the possible classes names in one shot, and check if all of them have their own documentation. An example, a custom doclet may use all the instances to retrieve the documentation of them in the plugin itself: this cannot be done with `getAllowedValuesForDescriptorHelp()`, because they are simple names.

Nevertheless, this is no high priority and we can open an issue to do it in the future...

Reply:
O, I just forgot that `getClassForPluginHelp` is exactly doing that for the `String` name. So they can be retrieved by iterating over `getAllowedValuesForDescriptorHelp` and using `getClassForPluginHelp` (although it may be inefficient, but good enough)